### PR TITLE
For next

### PR DESCRIPTION
**Pull request by hclee**

### CODE REVIEW

**Comment by namjaejeon** [comment]
Please explain the difference from the version 1 year ago in webos-next. and exfat2img ?

**Comment by hclee** [comment]
> Please explain the difference from the version 1 year ago in webos-next. and exfat2img ?

**Comment by hclee** [comment]
> Please explain the difference from the version 1 year ago in webos-next. and exfat2img ?

The difference is the way reclaimed files are created, and exfat2img not included. in webos-next orphaned clusters are connected to a newly created file(reclaimed file) if they are connected in FAT table, But in this patch set, if they are contiguous in allocation bitmap. This is the way Windows fsck does.

If exfat2img is needed, I have to refactor this patch set because fsck and exfat2img shares some code.

File: fsck/Makefile.am [hunk 1]
```diff
@@ -1,4 +1,4 @@
-AM_CFLAGS = -Wall -include $(top_builddir)/config.h -I$(top_srcdir)/include -fno-common
+AM_CFLAGS = -Wall -Wextra -include $(top_builddir)/config.h -I$(top_srcdir)/include -fno-common
```
[the remark this answers is not in the document]
Comment (hclee) [reply fsck/Makefile.am:1]: No, I think it is better the others have this option.

File: fsck/de_iter.c [hunk 2]
```diff
@@ -9,7 +9,10 @@
 
 #include "exfat_ondisk.h"
 #include "libexfat.h"
-#include "fsck.h"
+
```
[the remark this answers is not in the document]
Comment (hclee) [reply fsck/de_iter.c:12]: I have just run checkpatch without --strict. I will change this.

File: fsck/de_iter.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+/*
+ *  Copyright (C) 2021 Hyunchul Lee <hyc.lee@gmail.com>
+ */
+#ifndef _DE_ITER_H
+#define _DE_ITER_H
+
```
[the remark this answers is not in the document]
Comment (hclee) [reply fsck/de_iter.h:7]: Yes, I will move some code to libexfat and include exfat2img.

File: fsck/exfat_fs.c [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+ *   Copyright (C) 2021 Hyunchul Lee <hyc.lee@gmail.com>
+ */
+
+#include <errno.h>
+
+#include "exfat_ondisk.h"
+#include "libexfat.h"
+
+#include "inode.h"
+#include "exfat_fs.h"
```
[the remark this answers is not in the document]
Comment (hclee) [reply fsck/exfat_fs.c:12]: when I move some code to libexfat, I will change this name.

File: fsck/inode.c [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+		if (exfat->alloc_bitmap)
+			free(exfat->alloc_bitmap);
+		if (exfat->disk_bitmap)
+			free(exfat->disk_bitmap);
+		free(exfat);
+	}
+}
+
+struct exfat *exfat_alloc_exfat(struct exfat_blk_dev *blk_dev, struct pbr *bs)
+{
```
[the remark this answers is not in the document]
Comment (hclee) [reply fsck/inode.c:127]: I will refactor these.

File: fsck/lookup.c [hunk 6]
```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/*
+ *   Copyright (C) 2021 Hyunchul Lee <hyc.lee@gmail.com>
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+
```
[the remark this answers is not in the document]
Comment (hclee) [reply fsck/lookup.c:10]: de_iter.c is for reading directory entries when traveling files, and lookup.c is for find a specific directory entry with a given name. I will think about it

File: fsck/fsck.c [hunk 7]
```diff
@@ -720,33 +720,43 @@ static int read_file(struct exfat_de_iter *de_iter,
 	return ret;
 }
 
-static bool read_volume_label(struct exfat_de_iter *iter)
+static int read_volume_label(struct exfat *exfat)
```
[the remark this answers is not in the document]
Comment (hclee) [reply fsck/fsck.c:811]: Okay, I will append detailed description

File: fsck/fsck.c [hunk 8]
```diff
@@ -607,7 +607,7 @@ static int check_inode(struct exfat_de_iter *iter, struct exfat_inode *node)
 	return valid ? ret : -EINVAL;
 }
 
-static int read_file_dentries(struct exfat_de_iter *iter,
+static int read_file_dentry_set(struct exfat_de_iter *iter,
```
[the remark this answers is not in the document]
Comment (hclee) [reply fsck/fsck.c:671]: Sorry, I can't understand this. Can you elaborate this?

File: fsck/fsck.c [hunk 9]
```diff
@@ -954,24 +958,27 @@ static int read_children(struct exfat_fsck *fsck, struct exfat_inode *dir)
...
+			if (dir == exfat->root)
+				break;
+			/* fall through */
 		default:
 			if (IS_EXFAT_DELETED(dentry->type))
 				break;
-			exfat_err("unknown entry type. 0x%x\n", dentry->type);
-			ret = -EINVAL;
-			goto err;
+			if (repair_file_ask(de_iter, NULL, ER_DE_UNKNOWN,
```
[the remark this answers is not in the document]
Comment (hclee) [reply fsck/fsck.c:1066]: Yes, I remember Windows does. I will check it again

File: fsck/create.c [hunk 10]
```diff
@@ -47,7 +47,8 @@ static uint16_t calc_dentry_set_checksum(struct exfat_dentry *dset, int dcount)
 	return checksum;
 }
 
-static uint16_t calc_name_checksum(struct exfat *exfat, __le16 *name, int len)
+uint16_t exfat_calc_name_hash(struct exfat *exfat,
```
[the remark this answers is not in the document]
Comment (hclee) [reply fsck/create.c:50]: Okay, I will append detailed description.

File: fsck/fsck.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (namjaejeon) [reply fsck/fsck.c:671]: I need to find it again. Please update the patch except this comment.

**Comment by namjaejeon** [comment]
> The difference is the way reclaimed files are created,

How did you deal with orphan clusters before that? And what's problem if they are not move to reclaimed file?

> exfat2img not included

Please include it on this PR.

**Comment by hclee** [comment]
> > The difference is the way reclaimed files are created,
> 
> How did you deal with orphan clusters before that?

The previous version allocated orphan clusters which are chained in FAT table to a same reclaimed file,
But this version allocates orphan clusters which are contiguous in bitmap to a same reclaimed file.
This way Windows fsck does.

>And what's problem if they are not move to reclaimed file?
 
They are allocated, but there is no files connected with them.

> > exfat2img not included
> 
> Please include it on this PR.

Okay, I will.

**Comment by namjaejeon** [comment]
Is there any plan for next PR ?

**Comment by hclee** [comment]
I am going to do a pull request next week.

File: fsck/fsck.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (hclee) [reply fsck/fsck.c:1066]: Windows fsck also delete those.

**Comment by namjaejeon** [comment]
> The difference is the way reclaimed files are created,

@hclee Can you tell me what patch included this ?

**Comment by hclee** [comment]
> > The difference is the way reclaimed files are created,
> 
> @hclee Can you tell me what patch included this ?

You can find the patch below:
https://github.com/exfatprogs/exfatprogs/pull/199/commits/e7e39fd79b9d690b9bae784876ce0e29f888df2c

**Comment by namjaejeon** [comment]
> You can find the patch below:

Let's check if it is a fix that can't contribute whole repair patches for 1 year.

**Comment by hclee** [comment]
I have done a pull request for v2

**Comment by namjaejeon** [comment]
You have added case by case corruption image to exfatprogs. Is it possible to add fsck.exfat test using corruption images to Travis-CI script ?

**Comment by hclee** [comment]
> Reference



> You have added case by case corruption image to exfatprogs. Is it possible to add fsck.exfat test using corruption images to Travis-CI script ?

I think it is possible. I will do.

**Comment by hclee** [comment]
I have forced-pushed again, Because exfat2img has integer overflow problem in dump_bytes_to_stdout() when dumping file system to stdout.

File: lib/exfat_dir.c [hunk 11]
```diff
@@ -6,12 +6,24 @@
...
 #include "exfat_ondisk.h"
 #include "libexfat.h"
 #include "exfat_fs.h"
 #include "exfat_dir.h"
 
+static struct path_resolve_ctx path_resolve_ctx;
+
+#define fsck_err(parent, inode, fmt, ...)		\
+({							\
+		exfat_resolve_path_parent(&path_resolve_ctx,	\
```
Comment (namjaejeon) [inline lib/exfat_dir.c:21]: 2 tab ?

File: fsck/fsck.c [hunk 12]
```diff
@@ -1190,6 +1194,9 @@ static int exfat_root_dir_check(struct exfat *exfat)
 	exfat_debug("root directory: start cluster[0x%x] size[0x%" PRIx64 "]\n",
 		root->first_clus, root->size);
 
+	if (read_volume_label(exfat))
```
Comment (namjaejeon) [inline fsck/fsck.c:1205]: Can you explain why read_volume_label() move to here ? There is no reason in patch description.
And can you check patch description in other patches also ?

File: tests/loop_chain/config [hunk 13]
```diff
@@ -0,0 +1 @@
+#OPTS: -s
```
Comment (namjaejeon) [inline tests/loop_chain/config:1]: s/containa/contain/

File: lib/libexfat.c [hunk 14]
```diff
@@ -668,6 +662,21 @@ int exfat_get_next_clus(struct exfat *exfat, struct exfat_inode *node,
 	return 0;
 }
 
+int exfat_get_inode_next_clus(struct exfat *exfat, struct exfat_inode *node,
```
Comment (namjaejeon) [inline lib/libexfat.c:697]: You create this function in this patch. Why ? where is used ? there is no reason in patch description. and if you add new function in patch, next patch should use this.

File: include/exfat_dir.h [hunk 15]
```diff
@@ -37,10 +37,11 @@ struct exfat_lookup_filter {
 	struct {
 		struct exfat_dentry	*dentry_set;
 		int			dentry_count;
+		off_t			file_offset;
 		/* device offset where the dentry_set locates, or
 		 * the empty slot locates or EOF if not found.
 		 */
-		off_t			dentry_d_offset;
+		off_t			dev_offset;
```
Comment (namjaejeon) [inline include/exfat_dir.h:45]: Again and again, Why offset is stored to exfat_lookup_filter ?

File: lib/exfat_dir.c [hunk 16]
```diff
@@ -502,28 +502,43 @@ int exfat_lookup_file(struct exfat *exfat, struct exfat_inode *parent,
...
-		*time_ms = (tm.tm_sec & 1) * 100;
+	*checksum = ((*checksum << 15) | (*checksum >> 1)) + bytes[0];
+	*checksum = ((*checksum << 15) | (*checksum >> 1)) + bytes[1];
+
+	i = primary ? 4 : 2;
+	for (; i < sizeof(*dentry); i++)
+		*checksum = ((*checksum << 15) | (*checksum >> 1)) + bytes[i];
+}
+
+static uint16_t calc_dentry_set_checksum(struct exfat_dentry *dset, int dcount)
```
Comment (namjaejeon) [inline lib/exfat_dir.c:521]: So, Why do you add this functions ? Do you think that it is good to describe in patch description ?

**Comment by namjaejeon** [comment]
It seems that the patch description is still not well done. Also, I'm not sure if the patch order is correct.

**Comment by namjaejeon** [comment]
Once I have pushed this PR, Please send the new PR to exfat-next if you are missing something.

**Comment by hclee** [comment]
> Once I have pushed this PR, Please send the new PR to exfat-next if you are missing something.

I will the new PR to update licences tonight.

**Pull request merged**